### PR TITLE
Experimental interfaces and traits for `TensorNetwork`-derived types

### PR DESCRIPTION
**Pull request by mofeing**
This is a experiment to have a cleaner and more reusable interfaces between types in the `TensorNetwork` hierarchy of types.

# To do

- [x] Refactor `replace!`
- [x] Rethink about using `Core.kwcall` for keyword-dispatching
- [x] Write a post on docs about how this works internally
- [x] Remove calls to previous "dispatch system" (not really dispatch)
- [ ] Fix docstrings of `@kwdispatch`/`@kwmethod` methods

# Related links
- https://github.com/bcbi/ExtensibleUnions.jl
- https://github.com/jolin-io/WhereTraits.jl
- https://github.com/WschW/StructuralInheritance.jl
- https://github.com/mind6/Inherit.jl
- https://github.com/thautwarm/CanonicalTraits.jl
- https://github.com/Suzhou-Tongyuan/ObjectOriented.jl
- https://github.com/EdBarrancos/ObjectSystem.jl
- https://github.com/rjplevin/Classes.jl

### CODE REVIEW

**Comment by mofeing** [comment]
It's weird that the test fails when it works in my laptop...

**Comment by Todorbsc** [comment]
> It's weird that the test fails when it works in my laptop...

They are working in my laptop as well. It is something about the CI that we must check but, from my side, this is ready for merge :+1:

File: src/Quantum.jl [hunk 1]
```diff
@@ -1,159 +1,94 @@
...
-    if isempty(kwargs)
-        collect(keys(tn.sites))
-    elseif keys(kwargs) === (:at,)
-        findfirst(i -> i === kwargs[:at], tn.sites)
-    else
-        throw(MethodError(sites, (Quantum,), kwargs))
-    end
-end
+# outputs(q::Quantum) = sort!(collect(filter(!isdual, keys(q.sites))))
+@deprecate outputs(tn::AbstractQuantum) sites(tn; set=:outputs)
```
Comment (starsfordummies) [inline src/Quantum.jl:50]: I don't know why but (same as you yesterday) I don't get any warning about your deprecations here (though it seems to be calling the deprecate.. ) - did you force disable warns somewhere ?


```
julia> inputs(mps1)
Set{Site}()

julia> mps1 isa AbstractQuantum
ERROR: UndefVarError: `AbstractQuantum` not defined in `Main`
Suggestion: check for spelling errors or missing imports.
Stacktrace:
 [1] top-level scope
   @ REPL[13]:1

julia> mps1 isa Tenet.AbstractQuantum
true

julia> outputs(mps1)
Set{Site} with 5 elements:
  3
  5
  1
  2
  4

julia> ninputs(mps1)
0

julia> @which ninputs(mps1)
ninputs(tn::Tenet.AbstractQuantum)
     @ Tenet deprecated.jl:103
```

Comment (mofeing) [reply src/Quantum.jl:50]: Yeah, it's weird. This also happens to me but shouldn't be like this.

Comment (starsfordummies) [reply src/Quantum.jl:50]: ok, apparently you need to start julia with `julia --depwarn=yes` or [they don't show](https://invenia.github.io/blog/2022/06/17/deprecating-in-julia/#:~:text=Deprecation%20warnings%20are%20suppressed%20in,starting%20a%20new%20Julia%20session.)

meh 

```
julia> mps1 = rand(Chain, Open, State; n=5, d=2, χ=5 )
MPS (inputs=0, outputs=5)

julia> inputs(mps1)
┌ Warning: `inputs(tn::AbstractQuantum)` is deprecated, use `sites(tn; set = :inputs)` instead.
│   caller = top-level scope at REPL[5]:1
└ @ Core REPL[5]:1
Set{Site}()
```

Comment (mofeing) [reply src/Quantum.jl:50]: ahh that makes sense. I think that the Julia subprocesses spawned by `test` do have that flag activated.

Comment (jofrevalles) [reply src/Quantum.jl:50]: Thanks! @starsfordummies

**Comment by mofeing** [comment]
Finally I managed to reproduce the test fails in my laptop. Still don't know why it was working before :shrug:.

**Comment by Todorbsc** [comment]
> Finally I managed to reproduce the test fails in my laptop. Still don't know why it was working before 🤷.

What did you do to make them fail in your laptop?

**Comment by mofeing** [comment]
Nothing really. I think it was something related to some commits or some bug in the code. Still weird.

**Review by jofrevalles — APPROVED** [review]

**Review by starsfordummies — APPROVED** [review]

**Comment by mofeing** [comment]
@jofrevalles @Todorbsc @starsfordummies the only thing that bugs me is if we want all methods to be callable from subtypes; i.e. we want to be able to `push!`/`delete!` `Tensor`s from a MPS.

**Pull request merged**